### PR TITLE
refactor: remove user session singleton (part 6) - WPB-4920

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-4920" title="WPB-4920" target="_blank">BugWPB-4920</a>  [iOS] Crash when sending message in C1
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

## Issues

Often the app crashes shortly after logging out or the session expiring.

## Causes

The UI code depends heavily on the active user session, which is accessed via a singleton ZMUserSession.shared(). This is often safely unwrapped, but there are places where it is not, leading to a a crash. The most common point of crash is when the the static function ZMUser.selfUser() is accessed, which depends on the shared user session and implicitly unwraps its value.

In theory, the UI code should be running when there is no user session, but this can not be guaranteed, because the UI code may not have been torn down yet, or part of it may be retained and fired as a block in response to a notification or other observation mechanism.

## Solutions

Avoid all static access to the shared user session and self user. Instead, inject a non optional value from the top level and propagate it where it needs to go via dependency injection.

## Note

Any refinement to better model the user session and dependency structure isn't considered here. First we want to remove the static accessors and make it work. After we can polish it.

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 1]
```diff
@@ -53,14 +58,60 @@ public protocol UserSession: AnyObject {
 
     var appLockTimeout: UInt { get }
 
+    /// Whether a custom passcode has been set.
+
+    var isCustomPasscodeSet: Bool { get }
```
Comment (johnxnguyen) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift]: `isCustomAppLockPasscodeSet`

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 2]
```diff
@@ -53,14 +58,60 @@ public protocol UserSession: AnyObject {
...
+
+    /// Whether a custom passcode (rather a device passcode) should be used.
+
+    var requireCustomPasscode: Bool { get }
```
Comment (johnxnguyen) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift]: `requireCustomAppLockPasscode`

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 3]
```diff
@@ -53,14 +58,60 @@ public protocol UserSession: AnyObject {
...
+
     /// Whether the user should be notified of the app lock being disabled.
 
-    var shouldNotifyUserOfDisabledAppLock: Bool { get }
+    var shouldNotifyUserOfDisabledAppLock: Bool { get set }
+
+    /// Whether the user needs to be informed about configuration changes.
+
+    var needsToNotifyUser: Bool { get set }
```
Comment (johnxnguyen) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift]: Needs to notify user about...?

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 4]
```diff
@@ -53,14 +58,60 @@ public protocol UserSession: AnyObject {
...
+
+    /// Unlock the database using the given authentication context.
+
+    func unlockDatabase(with context: LAContext) throws
+
+    /// Open the app lock.
+
+    func open() throws
```
Comment (johnxnguyen) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift]: `openAppLock`

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 5]
```diff
@@ -53,14 +58,60 @@ public protocol UserSession: AnyObject {
...
+
+    /// Authenticate with device owner credentials (biometrics or passcode).
+    ///
+    /// - Parameters:
+    ///     - passcodePreference: Used to determine which type of passcode is used.
+    ///     - description: The message to dispaly in the authentication UI.
+    ///     - context: The context in which authentication happens.
+    ///     - callback: Invoked with the authentication result.
+
+    func evaluateAuthentication(
```
Comment (johnxnguyen) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift]: `evaluateAppLockAuthentication`

File: wire-ios/Wire-iOS Tests/App Lock/Tests/AppLockModuleRouterTests.swift
Comment (johnxnguyen) [inline wire-ios/Wire-iOS Tests/App Lock/Tests/AppLockModuleRouterTests.swift:1]: Why was this deleted?

File: wire-ios/Wire-iOS/Sources/UserInterface/App Lock/AppLockModule.Router.swift [hunk 6]
```diff
@@ -63,7 +68,7 @@ extension AppLockModule.Router: AppLockRouterPresenterInterface {
     private func presentInputPasscodeModule() {
         // TODO: [John] Clean this up.
         // TODO: [John] Inject these arguments.
-        let unlockViewController = UnlockViewController(selfUser: ZMUser.selfUser(), userSession: ZMUserSession.shared())
+        let unlockViewController = UnlockViewController(selfUser: ZMUser.selfUser(), userSession: userSession)
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/App Lock/AppLockModule.Router.swift:69]: You can get rid of these todos now.

File: wire-ios/Wire-iOS Tests/App Lock/Tests/AppLockModuleRouterTests.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/App Lock/Tests/AppLockModuleRouterTests.swift:1]: Because we didn't have any tests inside.

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 3 above)
Comment (typfel) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift]: `needsToNotifyUserOfAppLockConfiguration` ?

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 7]
```diff
@@ -91,6 +146,10 @@ public protocol UserSession: AnyObject {
         for: UserType
     ) -> NSObjectProtocol?
 
+    func addUserObservers(
```
Comment (typfel) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift]:
```suggestion
    func addUserObserver(
```

We are only adding one observer

File: wire-ios/Wire-iOS/Sources/UserInterface/CustomAppLock/Warning/AppLockChangeWarningViewController.swift [hunk 8]
```diff
@@ -136,11 +138,9 @@ final class AppLockChangeWarningViewController: UIViewController {
 
     @objc
     func confirmButtonTapped(sender: AnyObject?) {
-        if let session = ZMUserSession.shared() {
-            session.perform {
-                session.appLockController.needsToNotifyUser = false
+            userSession.perform {
+                self.userSession.shouldNotifyUserOfDisabledAppLock = false
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/CustomAppLock/Warning/AppLockChangeWarningViewController.swift]: Why do we now call `shouldNotifyUserOfDisabledAppLock` which modifies `AppLockControler.isActive`?

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 9]
```diff
@@ -166,6 +246,66 @@ extension ZMUserSession: UserSession {
...
 
+    public var requireCustomPasscode: Bool {
+        appLockController.requireCustomPasscode
+    }
+
+    public var isCustomPasscodeSet: Bool {
+        appLockController.isCustomPasscodeSet
+    }
+
+    public var shouldNotifyUserOfDisabledAppLock: Bool {
```
Comment (typfel) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift:253]: You are defining this computed property twice, it's also defined on line 309. Should maybe only keep that one? It's weird to have two setter which modify `appLockController.needsToNotifyUser`

**Review by typfel — CHANGES_REQUESTED** [review]

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 1 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift]: Done!

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift]: Done!

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift]: Done!

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 1 above)
Comment (caldrian) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift:63]: Btw, GitHub allows for "adding suggestion", which is just slightly more work for the commenter, but the PR author could easily apply the change.

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 4 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift]: Done!

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 5 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift]: Done!

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 9 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift:253]: I kept the one!

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 7 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift]: Done!

**Review by typfel — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/App Lock/AppLockModule.Router.swift (hunk 6 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/App Lock/AppLockModule.Router.swift:69]: I already took care of that @johnxnguyen

File: wire-ios-data-model/Source/Model/FeatureConfig/AppLock/AppLockController.swift [hunk 10]
```diff
@@ -116,7 +116,11 @@ public final class AppLockController: AppLockType {
 
     // MARK: - Life cycle
 
-    public init(userId: UUID, selfUser: ZMUser, legacyConfig: LegacyConfig? = nil) {
+    public init(
+        userId: UUID,
+        selfUser: ZMUser,
+        legacyConfig: LegacyConfig? = nil
```
Comment (markusfassbender) [inline wire-ios-data-model/Source/Model/FeatureConfig/AppLock/AppLockController.swift]:
```suggestion
        legacyConfig: LegacyConfig?
```
In general there is a chance one forgets to set arguments if they have default values. In this case the default value can be removed, because no caller uses it :)

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 11]
```diff
@@ -53,14 +58,60 @@ public protocol UserSession: AnyObject {
...
+
+    /// Authenticate with device owner credentials (biometrics or passcode).
+    ///
+    /// - Parameters:
+    ///     - passcodePreference: Used to determine which type of passcode is used.
+    ///     - description: The message to dispaly in the authentication UI.
+    ///     - context: The context in which authentication happens.
+    ///     - callback: Invoked with the authentication result.
+
+    func evaluateAppLockAuthentication(
```
Comment (markusfassbender) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift:93]: Do you have a convention about this empty line between documentation and definition or is it personal preference of each developer?

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 12]
```diff
@@ -91,6 +142,10 @@ public protocol UserSession: AnyObject {
         for: UserType
     ) -> NSObjectProtocol?
 
+    func addUserObserver(
+        _ observer: ZMUserObserver
+    ) -> NSObjectProtocol
```
Comment (markusfassbender) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift:147]: Do you have a convention about definition e.g. with a single parameter or is it personal preference of each developer?

File: wire-ios/Wire-iOS/Sources/Helpers/syncengine/ZMUserSession+Shared.swift [hunk 13]
```diff
@@ -19,7 +19,7 @@
 import WireSyncEngine
 
 extension ZMUserSession {
-
+    @available(*, deprecated, message: "This shared instance has been deprecated. Don't use it.")
     static func shared() -> ZMUserSession? {
         return SessionManager.shared?.activeUserSession
     }
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/Helpers/syncengine/ZMUserSession+Shared.swift:25]: ❤️

**Review by markusfassbender — APPROVED** [review]

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 11 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift:93]: I don't think we have such a convention atm. We can further discuss it and introduce one.

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 12 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift:147]: I think one parameter maybe it's a bit much but from 3/4 and above we try to split it up to multiple lines. With Xcode 15 is really easy to do with CTRL + M

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 11 above)
Comment (markusfassbender) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift:93]: OK I just think Apple does it without the line and I'm used to that. We can have chat somewhen
